### PR TITLE
Improved cli command line help - fixes #1005

### LINKED ISSUE #1005: Improve cli help documents and command line help
Appropriate the https://github.com/kubernetes/kubernetes/tree/master/pkg/kubectl/cmd/templates code so that we can do cool stuff like:

```golang
var (
	apply_long = templates.LongDesc(`
		Apply a configuration to a resource by filename or stdin.
		This resource will be created if it doesn't exist yet.
		To use 'apply', always create the resource initially with either 'apply' or 'create --save-config'.

		JSON and YAML formats are accepted.

		Alpha Disclaimer: the --prune functionality is not yet complete. Do not use unless you are aware of what the current state is. See https://issues.k8s.io/34274.`)

	apply_example = templates.Examples(`
		# Apply the configuration in pod.json to a pod.
		kubectl apply -f ./pod.json

		# Apply the JSON passed into stdin to a pod.
		cat pod.json | kubectl apply -f -`)
)

func NewCmdApply(f cmdutil.Factory, out io.Writer) *cobra.Command {
	var options ApplyOptions

	cmd := &cobra.Command{
		Use:     "apply -f FILENAME",
		Short:   "Apply a configuration to a resource by filename or stdin",
		Long:    apply_long,
		Example: apply_example,
		Run: func(cmd *cobra.Command, args []string) {
			cmdutil.CheckErr(validateArgs(cmd, args))
			cmdutil.CheckErr(cmdutil.ValidateOutputArgs(cmd))
			cmdutil.CheckErr(validatePruneAll(options.Prune, cmdutil.GetFlagBool(cmd, "all"), options.Selector))
			cmdutil.CheckErr(RunApply(f, cmd, out, &options))
		},
	}
}
```

### PR DESCRIPTION
**Pull request by robertojrojas**
Improved cli command line help by adding k8s templates, i18n, and examples. Fixes #1005

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2411)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
@k8s-bot ok to test

File: cmd/kops/create.go [hunk 1]
```diff
@@ -31,20 +31,37 @@ import (
...
 	resource.FilenameOptions
 }
 
+var (
+	create_long = templates.LongDesc(i18n.T(`
+		Create a resource by filename or stdin.`))
+
+	create_example = templates.Examples(i18n.T(`
+		# Create a cluster in AWS
+		kops create cluster --name=k8s.cluster.site \
```
Comment (chrislovecnm) [inline cmd/kops/create.go]: So this example should go Into kops create cluster command. This is an awesome example, but this command "create" accepts one of more yaml files ;)

File: cmd/kops/create.go [hunk 2]
```diff
@@ -31,20 +31,37 @@ import (
...
+		`))
+)
+
 func NewCmdCreate(f *util.Factory, out io.Writer) *cobra.Command {
 	options := &CreateOptions{}
 
 	cmd := &cobra.Command{
-		Use:   "create -f FILENAME",
-		Short: "Create a resource by filename or stdin.",
+		Use:     "create -f FILENAME",
```
Comment (chrislovecnm) [inline cmd/kops/create.go]: Filename(s) - can we do that??

File: cmd/kops/delete.go [hunk 3]
```diff
@@ -29,22 +29,36 @@ import (
...
 }
 
+var (
+	delete_long = templates.LongDesc(i18n.T(`
+	Delete clusters and instancegroups.
+	`))
+
+	delete_example = templates.Examples(i18n.T(`
+		# Delete a cluster in AWS.
+		kops delete cluster --name=k8s.cluster.site --state=s3://kops-state-1234
```
Comment (chrislovecnm) [inline cmd/kops/delete.go]: Should we add another example with --yes? The --yes flag immediately deletes a cluster.

Comment (chrislovecnm) [inline cmd/kops/delete.go]: So again this is an awesome example for delete cluster, and this is delete via a file. Can we do a delete with a file here and delete with a cluster name in delete cluster.

File: cmd/kops/delete.go [hunk 4]
```diff
@@ -29,22 +29,36 @@ import (
...
+)
+
 func NewCmdDelete(f *util.Factory, out io.Writer) *cobra.Command {
 	options := &DeleteOptions{}
 
 	cmd := &cobra.Command{
 		Use:        "delete -f FILENAME [--yes]",
-		Short:      "Delete clusters and instancegroups",
-		Long:       `Delete clusters and instancegroups`,
+		Short:      i18n.T("Delete clusters and instancegroups."),
```
Comment (chrislovecnm) [inline cmd/kops/delete.go]: UsIng a yaml file??? Not sure how to say that.

File: cmd/kops/edit.go [hunk 5]
```diff
@@ -21,17 +21,29 @@ import (
...
 
-This command changes the cloud specification in the registry.
+    This command changes the cloud specification in the registry.
 
-It does not update the cloud resources, to apply the changes use "kops update cluster".`,
+    It does not update the cloud resources, to apply the changes use "kops update cluster".`))
+
+	edit_example = templates.Examples(i18n.T(`
+		# Edit a cluster configuration in AWS.
+		kops edit cluster k8s.cluster.site --state=s3://kops-state-1234
```
Comment (chrislovecnm) [inline cmd/kops/edit.go]: What is a good domain? Use "kubernetes.io"??!

Comment (chrislovecnm) [inline cmd/kops/edit.go]: So this is the root command, which I do not think edits, you have to use edit cluster or edit ig.

File: cmd/kops/create.go (hunk 1 above)
Comment (chrislovecnm) [inline cmd/kops/create.go]: Let me take back what I said ;) This is awesome, we need example with a yaml file as well. Ping me an I can get you more info, as I want to be respectful of your time. Take a peek at the PR that was just put in about standards for writing cli docs ;)

File: cmd/kops/create.go (hunk 2 above)
Comment (justinsb) [inline cmd/kops/create.go]: This PR doesn't change that line => probably valid, but a separate PR

File: cmd/kops/edit.go (hunk 5 above)
Comment (justinsb) [inline cmd/kops/edit.go]: On the choice of domain, `example.com` is the traditional choice

File: docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,57 @@
+# Copyright 2017 The Kubernetes Authors.
```
Comment (justinsb) [inline docker/Dockerfile]: Do these three `docker/*` files belong in this PR?  Looks like they were added by mistake

File: docs/upgrade.md [hunk 7]
```diff
@@ -31,3 +31,15 @@ NOTE: rolling-update does not yet perform a real rolling update - it just shuts
  there will be downtime [Issue #37](https://github.com/kubernetes/kops/issues/37)
 We have implemented a new feature that does drain and validate nodes.  This feature is experimental, and you can use the new feature by setting `export KOPS_FEATURE_FLAGS="+DrainAndValidateRollingUpdate"`.
 
+### Terraform Users
+
+* `kops edit cluster $NAME`
+* set the KubernetesVersion to the target version (e.g. `v1.3.5`)
+* NOTE: The next 3 steps must all be ran in the same directory
+* `kops update cluster $NAME --target=terraform`
+* `terraform plan`
```
Comment (justinsb) [inline docs/upgrade.md]: I think technically we have to `cd out/terraform` after `kops update` and before `terraform plan`

**Comment by justinsb** [comment]
Looks great - thank you.  I think we need to remove those `docker/*` files unless you meant for them to be in this PR?

File: docs/upgrade.md (hunk 7 above)
Comment (chrislovecnm) [inline docs/upgrade.md]: @justinsb this file should not be in the PR either ~  @robertojrojas was having some git challenges today.  Oh I remember some rebase hell ...

Comment (robertojrojas) [inline docs/upgrade.md]: @justinsb @chrislovecnm I honestly don't know how these files ended up here. Like Chris said, I had some issues with Git rebase. Any suggestions on how to fix this?

File: cmd/kops/create.go (hunk 2 above)
Comment (geojaz) [inline cmd/kops/create.go]: I create -f all the time ;)

**Comment by chrislovecnm** [comment]
@geojaz asked him to squash and we can MERGE!!

**Comment by robertojrojas** [comment]
Closing due to Git issues.

**Pull request closed without merging**